drivers/adc/adm1177/iio_adm1177.c

```diff
@@ -0,0 +1,168 @@
+* IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+* LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************************************************************/
+#include "iio_adm1177.h"
+#include "adm1177.h"
+#include "no_os_error.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+#include "no_os_delay.h"
+
+static int adm1177_iio_read_raw(void *device, char *buf, uint32_t len,
+				const struct iio_ch_info *channel,
+				intptr_t priv)
+{
+	struct adm1177_iio_dev *iiodev = (struct adm1177_iio_dev *)device;
+	struct adm1177_dev *desc = (struct adm1177_dev *)iiodev->adm1177_dev;
+	int32_t ret, value;
+	uint16_t conv_voltage;
+	uint16_t conv_current;
+
+	switch(channel->type) {
+	case IIO_VOLTAGE:
+		ret = adm1177_read(desc, &conv_voltage, &conv_current);
+		if(ret)
+			return ret;
+
+		value = (int32_t)(conv_voltage);
+
+		break;
+	case IIO_CURRENT:
+		ret = adm1177_read(desc, &conv_voltage, &conv_current);
+		if(ret)
+			return ret;
+
+		value = (int32_t)(conv_current);
+
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return iio_format_value(buf, len, IIO_VAL_INT, 1, &value);
+}
+
+static struct iio_attribute adm1177_iio_voltage_attrs[] = {
+	{
+		.name = "raw",
+		.show = adm1177_iio_read_raw,
+		.store = NULL
+	},
+	END_ATTRIBUTES_ARRAY
+};
+
+static struct iio_attribute adm1177_iio_current_attrs[] = {
+	{
+		.name = "raw",
+		.show = adm1177_iio_read_raw,
+		.store = NULL
+	},
+	END_ATTRIBUTES_ARRAY
+};
+
+enum adm1177_chan {
+	VOLTAGE,
+	CURRENT,
+};
+
+static struct iio_channel adm1177_channels[] = {
+	[VOLTAGE] = {
+		.ch_type = IIO_VOLTAGE,
+		.indexed = true,
+		.channel = 0,
+		.attributes = adm1177_iio_voltage_attrs,
+		.address = ADM1177_VOLTAGE_EN,
+		.ch_out = true
+	},
+	[CURRENT] = {
+		.ch_type = IIO_CURRENT,
+		.indexed = true,
+		.channel = 1,
+		.attributes = adm1177_iio_current_attrs,
+		.address = ADM1177_CURRENT_EN,
+		.ch_out = true
+	}
+};
+
+static struct iio_device adm1177_iio_device = {
+	.num_ch = NO_OS_ARRAY_SIZE(adm1177_channels),
+	.channels = adm1177_channels,
+};
+
+int adm1177_iio_init(struct adm1177_iio_dev **iio_dev,
+		     struct adm1177_iio_init_param *init_param)
+{
+	int ret;
+	struct adm1177_iio_dev *desc;
+
+	desc = (struct adm1177_iio_dev *)no_os_calloc(1, sizeof(*desc));
+	if(!desc)
+		return desc;
+
+	desc->iio_dev = &adm1177_iio_device;
+
+	ret = adm1177_init(&desc->adm1177_dev, init_param->adm1177_initial);
+	if(ret)
+		goto error_desc;
+
+	ret = adm1177_write(desc->adm1177_dev, ADM1177_CMD_V_CONT |
```

Comment:
instead of making only the iio driver ready for curren/voltage reads why not also the main driver? (by moving thins operation in `adm1177_init`